### PR TITLE
Introduced the pi number, the euler's number, nan, angle

### PR DESCRIPTION
**Pull request by maxonfjvipon**
Ref: #597

### CODE REVIEW

**Comment by maxonfjvipon** [comment]
@yegor256 Questions and comments:  
1. Do we need `float.as-angle`?
1. If we want to have `angle.as-degrees` that converts radians to degrees I think we need `angle.as-float`
1. How much digits after dot are enough for `pi` and `e`?
1. Can we convert `nan` to `bytes` via `nan.as-bytes`?
1. Can we compare `nan` to something. [This](https://en.wikipedia.org/wiki/NaN) said `nan` can be compared but in all cases the result is `false` 
(except `nan != x` obviously)

**Comment by yegor256** [comment]
> Do we need float.as-angle?

No, we don't. `angle` must have a single argument `x`, which it must also decorate:

```
[x] > angle
  x > @
```

> If we want to have angle.as-degrees that converts radians to degrees I think we need angle.as-float

`angle` _is_ a `float`, no need to convert to it.

> How much digits after dot are enough for pi and e?

The most we can do.

> Can we convert nan to bytes via nan.as-bytes? What should be the result in that case?

I think this shouldn't be possible, we must return `[]`.

> Can we compare nan to something? [This](https://en.wikipedia.org/wiki/NaN) said nan can be compared but in all cases the result is false
(except nan != x obviously)

I think we should do as IEEE suggests.

**Comment by maxonfjvipon** [comment]
@yegor256 Please check last commit out. If it's good, I'll start to change some Java code.  
And one more thing about `nan`. If we try to `nan.add x` we get exception with "You can't add to NaN" message. `sub` "method" is implemented in pure eo and use `add` "method". For now if we try to `nan.sub x` with still get exception with "You can't `add` to NaN" message. Should we get "You can't `sub` to NaN" message?  
There are more "methods" that use other already existed "methods" like `div, neg, abs...`. It also related to them.

**Comment by yegor256** [comment]
@maxonfjvipon let's just return `You can't manipulate with NaN` error in `add`. The only method that should work in `nan` should be `eq`, I believe. `nan.eq nan` should be `TRUE`.

**Comment by yegor256** [comment]
@maxonfjvipon you know how to compare `nan` with `nan`, right? I would do it this way:

```
[] > nan
  [x] > eq
    x.<.eq 
      $.<
```

**Comment by maxonfjvipon** [comment]
@yegor256 yeah, I read about it in the paper before.

File: eo-runtime/src/main/eo/org/eolang/math/nan.eo [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    eq. > @
+      x.<
+      ^.<
+
+  # Tests that $ < x
+  [x] > less
+    FALSE > @
+
+  # Tests that $ ≤ x
+  [x] > leq
```
Comment (yegor256) [inline eo-runtime/src/main/eo/org/eolang/math/nan.eo:41]: @maxonfjvipon I don't think this is correct. I would expect `nan.leq nan` to be TRUE, while `nan.leq 10` to be FALSE.

File: eo-runtime/src/main/eo/org/eolang/math/nan.eo [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    FALSE > @
+
+  # Tests that $ > x
+  [x] > greater
```
Comment (yegor256) [inline eo-runtime/src/main/eo/org/eolang/math/nan.eo:45]: @maxonfjvipon why FALSE here? `nan.greater 5` should be TRUE, I believe, while `nan.greater nan` should be FALSE

File: eo-runtime/src/main/eo/org/eolang/math/nan.eo [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    FALSE > @
+
+  # Tests that $ ≥ x
+  [x] > geq
+    FALSE > @
```
Comment (yegor256) [inline eo-runtime/src/main/eo/org/eolang/math/nan.eo:50]: @maxonfjvipon also something is wrong here, `nan.geq nan` should be TRUE

**Comment by yegor256** [comment]
@maxonfjvipon see my comments above

**Comment by maxonfjvipon** [comment]
@yegor256 I was ruled by you message:
> I think we should do as IEEE suggests.

and by IEEE standart:
image

**Comment by yegor256** [comment]
@maxonfjvipon I vote for IEEE. But if we follow IEEE rules, `eq` must always return FALSE. In your pull request it doesn't.

**Comment by maxonfjvipon** [comment]
> @maxonfjvipon I vote for IEEE. But if we follow IEEE rules, `eq` must always return FALSE. In your pull request it doesn't.

@yegor256 will be fixed

**Comment by maxonfjvipon** [comment]
I think it's all good to write some Java

**Comment by yegor256** [comment]
@maxonfjvipon yes, looks good

**Comment by maxonfjvipon** [comment]
@yegor256 2 questions:  
1. Two tests when I try to compare result of `ctan` with `nan` were not passed. What am I doing wrong?
```
[] > ctan-zero
  eq. > @
    nan.<
    (angle 0.0).ctan.<

[] > ctan-pi
  eq. > @
    nan.<
    (angle pi).ctan.<
```
2. When I ran tests before on parsing step I got error when I tried to do this:

```
[] > tan-minus-pi-div-two
  eq. > @
    nan.<
    <.  # <----- unexpected token here
      tan.
        angle
          neg.
            pi div 2
```
If I understand right it's not allowed to do `<.`?

**Comment by yegor256** [comment]
@maxonfjvipon try this:

```
[] > ctan-zero
  eq. > @
    nan.<
    (angle 0.0).ctan.@.<

[] > ctan-pi
  eq. > @
    nan.<
    (angle pi).ctan.@.<
```

When you say `tan.<` it gives you the ID of the `tan` object, not of the result of its dataization. If you do `tan.@.<` you make it calculate the result and then take the ID.

**Comment by maxonfjvipon** [comment]
@yegor256 no result:(  
I tried this and what I got:

image

As you can see `nan.<` and `(angle pi).ctan.@.<` are different but there is no doubts that `(angle pi).ctan` returns `nan` because of `msg`.  

This is how `ctan` is defined:
```
# Cotangent of $
[] > ctan
  ^.sin > sine!
  if. > @
    sine.as-int.eq 0
    nan
    div.
      ^.cos
      sine
```
It seems everything should work fine... Any more ideas how to compare result of `ctan` with `nan`?  

Mb we can do something like this:
```
[] > ctan-pi-returns-nan
  eq. > @
    "You can't manipulate with NaN"
    ((angle pi).ctan.add 42).msg
```

But I think this is not really tests that "cotangent of `pi` equals to `nan`"

**Comment by yegor256** [comment]
@maxonfjvipon maybe it's a bug in the compiler, let me check...

**Comment by maxonfjvipon** [comment]
@yegor256 Hey, how's it going? Did you find out anything?

Mb while you're looking for what's the problem I could do something else?

**Comment by yegor256** [comment]
@maxonfjvipon this is the scope of work for the nearest future: https://github.com/yegor256/objectionary/issues You can pick any package/library and implement it in Java, as a new repository. You can use this one as a template, with all pom/maven configurations necessary: https://github.com/yegor256/eo-files

**Comment by yegor256** [comment]
@maxonfjvipon this seems to be a fundamental <del>bug</del> feature of EO. `tan.@.<` is the ID of the object that is bound to `@` attribute of `tan`. This object is `if`. So, you are getting the ID of `if`. Since EO is declarative, there is _no way_ to get the ID of the result of `tan` calculation using EO syntax: we don't deal with objects, we only deal with their behavior. We can compare whether the behavior of `x.tan` is similar to the behavior of `nan`, but we can't say that two objects are identical :( I'm thinking what could be a workaround.

**Comment by yegor256** [comment]
@maxonfjvipon the only option I see is this:

```
] > ctan-zero
  eq. > @
    nan.as-bytes
    (angle 0.0).ctan.as-bytes

[] > ctan-pi
  eq. > @
    nan.as-bytes
    (angle pi).ctan.as-bytes
```

Here, `as-bytes` of `nan` should return a unique combination of bits, which no other float number would have. Comparing it with the number returned by `tan` should give the result you are looking for.

**Comment by maxonfjvipon** [comment]
@yegor256 got you. I'll try out

**Comment by yegor256** [comment]
@maxonfjvipon how is it going?

**Comment by maxonfjvipon** [comment]
> @maxonfjvipon how is it going?

Hey! I'm sorry, I just got some university problems. There'll be a result by the end of the next week

**Comment by maxonfjvipon** [comment]
@yegor256 when we do `nan.as-bytes` we call `as-bytes` of float which `nan` decorates. Now `nan` decorates 3.14159265. Which means `nan.as-bytes` is the same `3.14159265.as-bytes`. And it means that `(angle 0.0).ctan.as-bytes` just doesn't equal to `3.14159265` but not `nan`. I hope you understand what I'm trying to say. `as-bytes` is still bad way to prove that something equals to `nan`.

**Comment by yegor256** [comment]
@maxonfjvipon why `nan` decorates `pi`? I think `nan.as-bytes` must return what [IEEE suggests](https://en.wikipedia.org/wiki/NaN). Not pi :)

**Comment by yegor256** [comment]
@maxonfjvipon 

> Mb after all nan should not decorate anything? We can just overwrite all float "methods" inside it.

Maybe so!

I suggest we represent `nan.as-bytes` the way IEEE suggests: https://en.wikipedia.org/wiki/IEEE_754-1985#NaN

**Comment by maxonfjvipon** [comment]
@yegor256 I made `nan` not decorate any float, seems good. `nan.as-bytes` returns `7F-C0-00-00` which is IEEE suggests. 
One more thing: we can do `nan.add 42`, but if we do `42.add nan` we get an exception. For some "methods" we could make a condition: if argument is `nan` return `nan`. But there are methods like `add` or `mul` that are not implemented in eo but in java directly. IEEE says that any operation with `nan` must return `nan`. Any suggestions?
image

**Comment by yegor256** [comment]
@maxonfjvipon I don't think we need this pull request here any more. Let's put this code into eo-math library.

**Comment by maxonfjvipon** [comment]
> @maxonfjvipon I don't think we need this pull request here any more. Let's put this code into eo-math library.

@yegor256 Done!

**Pull request closed without merging**